src/System.Text.Json/tests/JsonObjectTests.cs

```diff
@@ -0,0 +1,609 @@
+
+            Assert.IsType<JsonBoolean>(person1["is_married"]);
+            Assert.Equal(false, person1["is_married"] as JsonBoolean);
+
+            var person2 = new JsonObject
+            {
+                { "name", "Bob" },
+                { "age", 33 },
+                { "is_married", true }
+            };
+
+            // Copy property from another JsonObject
+            person1["age"] = person2["age"];
+
+            Assert.IsType<JsonNumber>(person1["age"]);
+            Assert.Equal(33, person1["age"] as JsonNumber);
+
+            // Copy property of different typoe
+            person1["name"] = person2["name"];
+        }
+
+        [Fact]
+        public static void TestModifyingJsonObjectKeyRemoveAdd()
+        {
+            JsonObject manager = EmployeesDatabase.GetManager();
+            JsonObject reportingEmployees = manager.GetJsonObjectProperty("reporting employees");
+
+            static void ModifyProperty(JsonObject jsonObject, string previousName, string newName)
+            {
+                JsonNode previousValue = jsonObject[previousName];
+                jsonObject.Remove(previousName);
+                jsonObject.Add(newName, previousValue);
+            }
+
+            string previousName = "software developers";
+            string newName = "software engineers";
+
+            Assert.True(reportingEmployees.ContainsProperty(previousName));
+            JsonNode previousValue = reportingEmployees[previousName];
+
+            ModifyProperty(reportingEmployees, previousName, newName);
+
+            Assert.False(reportingEmployees.ContainsProperty(previousName));
+            Assert.True(reportingEmployees.ContainsProperty(newName));
+            Assert.Equal(previousValue, reportingEmployees[newName]);
+        }
+
+        [Fact]
+        public static void TestModifyingJsonObjectKeyModifyMethod()
+        {
+            JsonObject manager = EmployeesDatabase.GetManager();
+            JsonObject reportingEmployees = manager.GetJsonObjectProperty("reporting employees");
+
+            Assert.True(reportingEmployees.ContainsProperty("software developers"));
+            JsonNode previousValue = reportingEmployees["software developers"];
+
+            reportingEmployees.ModifyPropertyName("software developers", "software engineers");
+
+            Assert.False(reportingEmployees.ContainsProperty("software developers"));
+            Assert.True(reportingEmployees.ContainsProperty("software engineers"));
+            Assert.Equal(previousValue, reportingEmployees["software engineers"]);
+        }
+
+        [Fact]
+        public static void TestAccesingNestedJsonObjectCastWithAs()
+        {
+            JsonObject manager = EmployeesDatabase.GetManager();
+
+            // Should not throw any exceptions:
+
+            var reportingEmployees = manager["reporting employees"] as JsonObject;
+            if (reportingEmployees == null)
+                throw new InvalidCastException();
```

Comment:
Don't throw in tests like this. Instead, assert the test condition is false.

Reply:
Yes, sure.

Reply:
Or just hard-cast and let the runtime throw if it's the wrong type.

Reply:
The quick rule: The test code should "throw" if the test itself is broken... mainly for helper routines that receive bad input, or Theory methods that receive incorrect input.  It should Assert that the product is behaving correctly.

In this case, it could go either way.  If the test is legitimately testing that the object returned is not null and is the correct type, then

```C#
object reportingEmployees = manager["reporting employees"];
Assert.NotNull(reportingEmployees);
Assert.IsType<JsonObject>(reportingEmployees);
```

Note that it only uses the indexer once, because using the indexer twice is actually a different test... you want to perform all the tests that are relevant on the same return value.

On the other hand, if some other test has already confirmed the "JsonObject is returned when that's the right answer" behavior, then this test could just assume the product behaves and only Assert the parts it's actually uniquely testing.